src/utils/batch-operate/SessionQueue.ts

```diff
@@ -0,0 +1,153 @@
+
+	private client: AirtopClient;
+
+	constructor({
+		maxConcurrentSessions,
+		runEmitter,
+		maxWindowsPerSession,
+		initialBatches,
+		operation,
+		client,
+		sessionConfig,
+		onError,
+	}: {
+		maxConcurrentSessions: number;
+		runEmitter: EventEmitter;
+		maxWindowsPerSession: number;
+		initialBatches: BatchOperationUrl[][];
+		operation: (input: BatchOperationInput) => Promise<BatchOperationResponse>;
+		client: AirtopClient;
+		sessionConfig?: AirtopSessionConfigV1;
+		onError?: (error: Error | string) => void;
+	}) {
+		if (
+			!Number.isInteger(maxConcurrentSessions) ||
+			maxConcurrentSessions <= 0
+		) {
+			throw new Error("maxConcurrentSessions must be a positive integer");
+		}
+
+		this.maxConcurrentSessions = maxConcurrentSessions;
+		this.runEmitter = runEmitter;
+		this.maxWindowsPerSession = maxWindowsPerSession;
+		this.sessionConfig = sessionConfig;
+		this.initialBatches = initialBatches;
+		this.operation = operation;
+		this.onError = onError;
+		this.processingPromise = null;
+
+		this.client = client;
+	}
+
+	public async addUrlsToBatchQueue(
+		newBatch: BatchOperationUrl[],
+	): Promise<void> {
+		// Distribute new URLs across batches
+		const newBatches = distributeUrlsToBatches(
+			newBatch,
+			this.maxConcurrentSessions,
+		);
+
+		this.client.log(
+			`Adding new batches to queue: ${JSON.stringify(newBatches)}`,
+		);
+		this.batchQueue.push(...newBatches);
+
+		// Update existing processing promise
+		this.processingPromise = this.processPendingBatches();
+	}
+
+	public async processInitialBatches(): Promise<void> {
+		this.batchQueue = [...this.initialBatches];
+		this.processingPromise = this.processPendingBatches();
+		await this.processingPromise;
+	}
+
+	public async waitForProcessingToComplete(): Promise<void> {
+		await this.processingPromise;
+	}
+
+	private async processPendingBatches(): Promise<void> {
+		while (this.batchQueue.length > 0) {
+			if (this.activePromises.length < this.maxConcurrentSessions) {
+				const batch = this.batchQueue.shift();
+				if (!batch || batch.length === 0) break;
+
+				const promise = (async () => {
+					let sessionId: string | undefined;
+					try {
+						const { data: session } = await this.client.sessions.create({
+							configuration: this.sessionConfig,
+						});
+						sessionId = session.id;
+
+						const queue = new WindowQueue(
+							this.maxWindowsPerSession,
+							this.runEmitter,
+							sessionId,
+							this.client,
+							this.operation,
+							this.onError,
+						);
+						await queue.processInBatches(batch);
+					} catch (error) {
+						if (this.onError) {
+							this.onError(
+								error instanceof Error ? error.message : String(error),
+							);
+						} else {
+							this.client.error(
+								error instanceof Error ? error.message : String(error),
+							);
+						}
+					} finally {
+						if (sessionId) {
+							await this.client.sessions.terminate(sessionId);
```

Comment:
Good call out, updated the code to use a pool of sessions so we don't create new ones unnecessarily.  We'll still create a fresh window every time though since there's less overhead and I could see reusing windows being prone to errors